### PR TITLE
Quilt Registry Builder

### PR DESCRIPTION
**Pull request by Leo40Git**
Add registry builders!
- The base `QuiltRegistryBuilder` class
- Built-in (code-driven) registries: `QuiltBuiltinRegistryBuilder`
- Dynamic (data-driven) registries: `QuiltDynamicRegistryBuilder`
  - Registers into the `DynamicRegistryManager`!

### TODO
- [ ] Document dynamic registries more!!!
- [ ] Add dynamic registry testmod

### CODE REVIEW

**Review by TheGlitch76 — APPROVED** [review]

**Review by EnnuiL — APPROVED** [review]
I like this!

**Review by PotatoPresident — APPROVED** [review]

File: library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltRegistryBuilder.java [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+	 * Note that this method does <em>not</em> handle registering the registry into the
+	 * {@link net.minecraft.util.registry.DynamicRegistryManager DynamicRegistryManager}.
+	 * You will have to do this manually.
+	 *
+	 * @return this builder
+	 *
+	 * @throws IllegalArgumentException if this builder is configured with a {@linkplain #syncBehavior(SyncBehavior) synchronization behavior} that is not {@link SyncBehavior#SKIPPED}
+	 */
+	@Contract("-> this")
+	public @NotNull QuiltRegistryBuilder<T> dynamic() {
```
Comment (LambdAurora) [inline library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltRegistryBuilder.java]: So, I looked a bit into dynamic registries and to make this fully working we would need:
 - register the registry into [`BuiltinRegistries.REGISTRIES`](https://maven.quiltmc.org/repository/release/org/quiltmc/quilt-mappings/1.19.2+build.12/quilt-mappings-1.19.2+build.12-javadoc.jar/net/minecraft/util/registry/BuiltinRegistries.html#REGISTRIES)
 - ask for a codec and a network codec (which may be the same), and put those into [`DynamicRegistryManager.INFOS`](https://maven.quiltmc.org/repository/release/org/quiltmc/quilt-mappings/1.19.2+build.12/quilt-mappings-1.19.2+build.12-javadoc.jar/net/minecraft/util/registry/DynamicRegistryManager.html#INFOS)
 
 And I believe that's it.

**Review by LambdAurora — COMMENTED** [review]
Would be neat to also have some kind of `withBootstrap` method which takes a `Consumer<Registry>` so you can register some default values as soon as the registry is built, similar to some of Vanilla registries.

File: library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltRegistryBuilder.java (hunk 1 above)
Comment (Leo40Git) [reply library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltRegistryBuilder.java]: I think I'd prefer splitting the class into a `QuiltBuiltinRegistryBuilder` and a `QuiltDynamicRegistryBuilder`.

Comment (LambdAurora) [reply library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltRegistryBuilder.java]: Sure, and have a base builder class to extend, sounds fine.

**Comment by Leo40Git** [comment]
Marking as WIP/draft until 1) dynamic registry testmod is done and 2) dynamic registry JSON path is cleared up.

File: library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltDynamicRegistryBuilder.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+			if (supplier.isInitialized()) {
+				throw new IllegalStateException("Built-in DynamicRegistryManager was already initialized!");
+			}
+		}
+
+		BuiltinRegistries.register((Registry<Registry<Object>>) BuiltinRegistries.REGISTRIES, this.key.getValue(), (Registry<Object>) registry);
+
+		var infos = DynamicRegistryManager.INFOS;
+
+		infos.put(this.key, new DynamicRegistryManager.Info<>(this.key, this.entryCodec, this.networkEntryCodec));
```
Comment (Leo40Git) [inline library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltDynamicRegistryBuilder.java]:
```suggestion
		DynamicRegistryManager.INFOS.put(this.key, new DynamicRegistryManager.Info<>(this.key, this.entryCodec, this.networkEntryCodec));
```

**Comment by Leo40Git** [comment]
Undrafting, but marking as `s: not working` due to the missing testmod (which will be worked on soon-ish!)

Dynamic registries need more documentation, as well.

**Review by LambdAurora — APPROVED** [review]

File: library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltRegistryBuilder.java [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+	 *     }
+	 * }
+	 * </code></pre>
+	 * Then, set this to call that {@code getRegistryHolder()} method:
+	 * <pre><code>
+	 * public class YourCoolModRegistries {
+	 *     public static final RegistryKey<Registry<Thing>> THING_KEY = RegistryKey.ofRegistry(YourCoolMod.id("thing"));
+	 *
+	 *     public static final SimpleRegistry<Thing> THING =
+	 *         new QuiltRegistryBuilder(THING_KEY)
```
Comment (Leo40Git) [inline library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltRegistryBuilder.java:182]:
```suggestion
	 *         QuiltRegistryBuilder.builtin(THING_KEY)
```

File: library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltRegistryBuilder.java [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+		return this.withLifecycle(Lifecycle.deprecated(since));
+	}
+
+	/**
+	 * Sets the custom {@link Holder} provider of this registry.
+	 * <p>
+	 * This should be used if your registry entry needs to store its own holder. Add the following to your registry entry class:
+	 * <pre><code>
+	 * public class Thing {
+	 *     private final Holder<Thing> registryHolder = YourCoolModRegistries.THINGS.createIntrusiveHolder(this);
```
Comment (Leo40Git) [inline library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltRegistryBuilder.java:169]:
```suggestion
	 *     private final Holder<Thing> registryHolder = YourCoolModRegistries.THING.createIntrusiveHolder(this);
```

**Comment by EnnuiL** [comment]
Partially superseded by PR #271, but not to a point where this PR becomes completely useless! In fact, it would be a great complement to that PR, and yeah, it no longer has to worry about the whole dynamic registry dilemma

**Pull request closed without merging**